pyk/src/pyk/kcfg/kcfg.py

```diff
@@ -244,6 +244,41 @@ def replace_target(self, node: KCFG.Node) -> KCFG.Edge:
             assert node.id == self.target.id
             return KCFG.Edge(self.source, node, self.depth, self.rules)
 
+    @final
+    @dataclass(frozen=True)
+    class MergedEdge(EdgeLike):
+        """Merged edge is a collection of edges that have been merged into a single edge."""
+
+        source: KCFG.Node
+        target: KCFG.Node
+        edges: tuple[KCFG.Edge, ...]
+
+        def to_dict(self) -> dict[str, Any]:
+            return {
+                'source': self.source.id,
+                'target': self.target.id,
+                'edges': [edge.to_dict() for edge in self.edges],
+            }
+
+        @staticmethod
+        def from_dict(dct: dict[str, Any], nodes: Mapping[int, KCFG.Node]) -> KCFG.Successor:
+            return KCFG.MergedEdge(
+                nodes[dct['source']],
+                nodes[dct['target']],
+                tuple(KCFG.Edge.from_dict(edge, nodes) for edge in dct['edges']),
+            )
+
+        def replace_source(self, node: KCFG.Node) -> KCFG.Successor:
```

Comment:
Does replacing the source (target) of such a merged edge not have an implication on the source (target) of the contained edges?

Reply:
Are we even allowed to do this, given the semantics of merged edges?

Reply:
Firstly, `Successor` has abstract methods `replace_source` and `replace_target`, so I had to implement them in the `MergedEdge`. Although it may seem a bit odd semantically, it does provide some syntactic and operational convenience.

Secondly, there is a semantic relationship here, but we cannot verify this implication without an interpreter (SymbolicCTerm). And from what I understand, not all implications can be solved in Z3.

Reply:
The `MultiEdge` should function just as a normal edge, just that it carries on it extra data, specifically:

- A `list` of depths/rules/constraints attached to it, indicating the different paths to take to get between the source and target.

I don't think `replace_[source|target]` really need to be treated specially here. They are maintenance functions for updating the nodes in an edge. They are used, for example, when calling `simplify` on all the nodes in the KCFG and we need to update the references on the edges. They exist just because we still store actual `Node` on the edges in the graph, rather than just references to a node dictionary.

Reply:
So, I think that I can resolve this conversation